File: algorithms/medium/arithmetic_slices.ex
defmodule Solution do
  @spec number_of_arithmetic_slices(nums :: [integer]) :: integer
  def number_of_arithmetic_slices(nums) do
    nums
    |> Enum.with_index()
    |> aux(0, 0)
  end

  def aux([], _ct, ans), do: ans

  def aux([_], _ct, ans), do: ans

  def aux([_, _], _ct, ans), do: ans

  def aux([{a, i}, {b, j}, {c, k} | rest], ct, ans) do
    if a - b == b - c do
      aux([{b, j}, {c, k} | rest], ct + 1, ans + ct + 1)
    else
      aux([{b, j}, {c, k} | rest], 0, ans)
    end
  end
end

File: algorithms/medium/arithmetic_slices.rs
fn ct(left: usize, right: usize) -> i32 {
    if right < left + 2 {
        return 0;
    }
    let n = right - left + 1 - 2;
    ((1 + n) * n / 2) as i32
}

impl Solution {
    pub fn number_of_arithmetic_slices(nums: Vec<i32>) -> i32 {
        if nums.len() < 3 {
            return 0;
        }
        let mut ans = 0;
        let mut left = 0;
        let mut right = 1;
        for i in 2..nums.len() {
            if nums[i] - nums[i-1] == nums[i-1] - nums[i-2] {
                right = i;
            } else {
                ans += ct(left, right);
                left = i - 1;
            }
        }

        ans += ct(left, right);
        ans
    }
}
